notebooks/corridor/corridor_bucharest_river-valley.qmd

```diff
@@ -0,0 +1,299 @@
+end_points_side_2 <- get_nearest_nodes(network_classified, end_points, "side_2")
+```
+
+We then define the initial guesses for the corridor edges by splitting the valley boundary (as determined from the DEM) in two segments using the waterway as dividing line:
+
+```{r}
+# split valley boundary in two segments
+edges_initial <- split_river_sides(valley, waterway) |>
+    st_boundary() |>
+    st_difference(waterway)
+
+# find out which segment is on which side of the river
+mask <- st_intersects(river_sides, edges_initial, sparse = FALSE)
+edge_side_1 <- edges_initial[mask[1,]]
+edge_side_2 <- edges_initial[mask[2,]]
+names(edge_side_1) <- "side_1"
+names(edge_side_2) <- "side_2"
+```
+
+We add weights to the network edges. These weights are used at a later stage to find the optimal paths connecting the end points on each side of the waterway. The weights account for the following elements:
+- the lenght of the edge: we want to favour shorter paths over longer ones;
+- the distance between the edge and the initial guess of the corridor boundary: we look for paths that lay as close as possible to the initial guess of the corridor.
+- whether an edge is contained within the initial guess of the corridor: we add a "penalty" cost to edges that fall within the corridor.
+The weight assigned to each edge is a linear combination of these components - coefficients so far are arbitrarily defined!
+
+```{r}
+# add edge weights to the network, accounting for the combined effect of the
+# edge length, distance from a give geometry, intersection with a buffer region
+add_weights <- function(net, geom, buffer){
+    edges <- get_edges(net)
+    distances <- st_distance(edges, geom, which = "Euclidean")
+    # the following is very slow, use approximate (faster) approach
+    # is_inside <- edges |>
+        # st_covered_by(buffer, sparse = FALSE)
+    is_inside <- edges |>
+        st_centroid() |>
+        st_intersects(buffer, sparse = FALSE)
+    # convert this to numeric values, and assign it a fictitious
+    # unit in order to be able to combine it with the other commpondents
+    # of the weight
+    is_inside <- as.numeric(is_inside) |> set_units(m)
+    name <- sprintf("weight_%s", names(geom))
+    net |>
+        activate("edges") |>
+        mutate("{name}" := 1. * edge_length() + 1. * distances + 1000. * is_inside)
+}
+
+network_weights <- network_classified |>
+    add_weights(edge_side_1, valley) |>
+    add_weights(edge_side_2, valley)
+```
+
+As a final network preparation step, we drop the edges that cross the waterway. This is to avoid that the shortest-path solutions cross the river:
+
+```{r}
+not_intersects <- function(x, y) !st_intersects(x, y)
+
+network_split <- network_weights  |>
+    activate("edges") |>
+    st_filter(waterway, .predicate = not_intersects) |>
+    activate("nodes") |>
+    filter(!node_is_isolated())
+```
+
+TODO: is enough to drop isolated nodes, or should we check for the largest group components?
```

Comment:
Indeed checking for the largest component should be more reliable as the removal of crossing segment may create more then two connected components besides isolated nodes.

Reply:
Good point, we should definitely keep the two largest components in the final version in the package!